### DISCUSSION TITLE
Solved: How to make two polygons align as if they're in the same coordinate system

### DISCUSSION THREAD

**Discussion by ReonKovalchuk** [fabricjs/fabric.js#9819] [Q&A] [answered]
I have two different situations, that basically come to one question: _how do I make two polygons behave as if their points are in the same coordinate system?_ 

Clarification:
**Situation 1:** two polygons, one is offset of the other. When I transform points of polygon to acute angles, offset polygon cannot be centered in `centerPoint = mainPolygon.getCenterPoint()`, it comes out of alignment.
Here's [jsFiddle](https://jsfiddle.net/micro_lion/v0qj5rgk/50/) and stackOverflow [question](https://stackoverflow.com/questions/78353774/centering-offset-polygon-inside-main) with some additional info

**Situation 2:** two polygons again. One is just a shape, that should be in a particular place of the other, like so
![Снимок экрана 2024-04-24 в 13 30 21](https://github.com/fabricjs/fabric.js/assets/105048534/9f47403e-d300-4851-a090-b66259245da8)

In both cases, all I have are coordinates of polygons points, that would give me a correct picture, if I would've drawn it on paper. 
Do i need to put them in a group and transform one of the polygon's coordinates?

Thank you for your help and your work on this project!

**EDIT**
Figured it out!
used this to translate secondary polygon's points into the same space as main polygon's points.

```
const centerPoint = polygon.getCenterPoint();
secondaryPoly.setPositionByOrigin(centerPoint, "center", "center");
const transformedPoints = offsetPoints.map((p) => {
    p.x = p.x - polygon.pathOffset.x + secondaryPoly.pathOffset.x;
    p.y = p.y - polygon.pathOffset.y + secondaryPoly.pathOffset.y;
    return p;
  });
secondaryPoly.points = transformedPoints;
```
And if the polygon is rotated

```
const centerPoint = polygon.getCenterPoint();
secondaryPoly.setPositionByOrigin(centerPoint, "center", "center");
const transformedPoints = offsetPoints.map((p) => {
    p.x = p.x - polygon.pathOffset.x + encumberance.pathOffset.x;
    p.y = p.y - polygon.pathOffset.y + encumberance.pathOffset.y;
    //account for angle and rotate the point
    const newPoint = fabric.util.rotatePoint(
      p,
      secondaryPoly.getCenterPoint(),
      fabric.util.degreesToRadians(polygon.angle)
    );
return newPoint;
  });
secondaryPoly.points = transformedPoints;
```

**Comment by ShaMan123** [comment, marked as answer]
Nice job
FYI we exposed a number of utils that make it a bit easier to use... but still there is a long way to go
Checkout `sendPointToPlane/sendVectorToPlane/sendObjectToPlane`
In your case:
```
const vertexAInPlaneOfVertexB = fabric.util.sendPointToPlane(
  vertexA, 
  fabric.util.multiplyTransformMatrices(
      polyA.calcTransformMatrix(), 
      fabric.util.createTranslateMatrix(-polyA.pathOffset.x, -polyA.pathOffset.y)
   ),
  fabric.util.multiplyTransformMatrices(
      polyB.calcTransformMatrix(), 
      fabric.util.createTranslateMatrix(-polyB.pathOffset.x, -polyB.pathOffset.y)
   ),
)
```

This solution is 100% safe and is suited for nested objects as well where as your solution will not work under groups.
Happy linear algebra!

**Reply by ReonKovalchuk** [reply]
I was sure there is a "proper" way)) thank you for providing the extended answer

**Reply by ShaMan123** [reply]
sure! don't hesitate to ask, sorry for th late reply

**Reply by ShaMan123** [reply]
BTW I use these 3 utils everywhere for everything
I have a 4th util that is as powerful as all of them combined which does a basis change
I plan to expose it
